Element tracking

robin-marr-ft opened the pull request:
/cc @AlbertoElias @kavanagh

kaelig commented on test/viewport.test.js:
```diff
@@ -54,4 +84,19 @@ describe('o-viewport', function() {
+		});
+
+		setTimeout(function () {
+			expect(true).to.be.true;
```
That's a funny one

kaelig commented on test/viewport.test.js:
```diff
@@ -54,4 +84,19 @@ describe('o-viewport', function() {
+	it('should stop listening to scroll event', function(done) {
+		oViewport.stopListeningTo('scroll');
+		document.body.addEventListener('oViewport.scroll', function(ev) {
+			expect(false).to.be.true;
```
![quote-william-s -burroughs-nothing-is-true-everything-is-permitted-124339_1](https://cloud.githubusercontent.com/assets/85783/6806103/aca69098-d23e-11e4-8c7e-81a4c5560005.png)

robin-marr-ft replied:
Forgot I was supposed to look up how to do that properly, I've updated it
now.


On 24 March 2015 at 15:58, Kaelig <notifications@github.com> wrote:

> In test/viewport.test.js
> <https://github.com/Financial-Times/o-viewport/pull/17#discussion_r27042323>
> :
>
> > @@ -54,4 +84,19 @@ describe('o-viewport', function() {
> >  	it('should get the orientation of the viewport', function() {
> >  		expect(oViewport.getOrientation() === 'portrait' || oViewport.getOrientation() === 'landscape').to.be(true);
> >  	});
> > +
> > +	it('should stop listening to scroll event', function(done) {
> > +		oViewport.stopListeningTo('scroll');
> > +		document.body.addEventListener('oViewport.scroll', function(ev) {
> > +			expect(false).to.be.true;
>
> [image: quote-william-s
> -burroughs-nothing-is-true-everything-is-permitted-124339_1]
> <https://cloud.githubusercontent.com/assets/85783/6806103/aca69098-d23e-11e4-8c7e-81a4c5560005.png>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Financial-Times/o-viewport/pull/17/files#r27042323>.
>

-- 

------------------------------
This email was sent by a company owned by Pearson plc, registered office at 
80 Strand, London WC2R 0RL.  Registered in England and Wales with company 
number 53723.

AlbertoElias commented on .gitignore:
```diff
@@ -2,5 +2,7 @@
 .sass-cache/
 bower_components/
 node_modules/
+demos/
```
Why are you ignoring demos? You should just be ignoring demos/local

robin-marr-ft replied:
Because when I run obt demo --local I get the below tree, this gitignore
config will ignore everything in red:

demos
├── element-tracking.html
├── local
│   ├── element-tracking.css
│   ├── element-tracking.html
│   └── element-tracking.js
└── src

    ├── config.json

    ├── element-tracking.js

    ├── element-tracking.mustache

    └── element-tracking.scss

On 25 March 2015 at 10:46, Alberto Elias <notifications@github.com> wrote:

> In .gitignore
> <https://github.com/Financial-Times/o-viewport/pull/17#discussion_r27109867>
> :
>
> > @@ -2,5 +2,7 @@
> >  .sass-cache/
> >  bower_components/
> >  node_modules/
> > +demos/
>
> Why are you ignoring demos? You should just be ignoring demos/local
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Financial-Times/o-viewport/pull/17/files#r27109867>.
>

-- 

------------------------------
This email was sent by a company owned by Pearson plc, registered office at 
80 Strand, London WC2R 0RL.  Registered in England and Wales with company 
number 53723.

AlbertoElias replied:
You're supposed to get everything that's not in local

robin-marr-ft replied:
I see, sorry, I was told not to commit the files in root, I've fixed this
now.

On 25 March 2015 at 12:15, Alberto Elias <notifications@github.com> wrote:

> In .gitignore
> <https://github.com/Financial-Times/o-viewport/pull/17#discussion_r27114961>
> :
>
> > @@ -2,5 +2,7 @@
> >  .sass-cache/
> >  bower_components/
> >  node_modules/
> > +demos/
>
> You're supposed to get everything that's not in local
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Financial-Times/o-viewport/pull/17/files#r27114961>.
>

-- 

------------------------------
This email was sent by a company owned by Pearson plc, registered office at 
80 Strand, London WC2R 0RL.  Registered in England and Wales with company 
number 53723.

commuterjoy commented:
out of interest, what's this for?

robin-marr-ft commented:
Hi Matt,

We found Paul and I were duplicating effort, o-ads needs to track elements
for lazy loading, sticky and refresh in view so we decided to move the
functionality into an origami module so it can be shared, after discussing
with the origami team we decided to add it to o-viewport

Cheers
Rob

On 25 March 2015 at 10:52, Matt Chadburn <notifications@github.com> wrote:

> out of interest, what's this for?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Financial-Times/o-viewport/pull/17#issuecomment-85981170>
> .
>

-- 

------------------------------
This email was sent by a company owned by Pearson plc, registered office at 
80 Strand, London WC2R 0RL.  Registered in England and Wales with company 
number 53723.

wheresrhys commented:
My gut feeling is that this should have oViewport as a dependency and be its own module. It seems like a massive departure from the original intention of the module, which was a minimal utility to discourage performance bottlenecks and normalise a few bowser quirks.

The code looks good, but I really don't think this is the place for it

kavanagh commented:
Agree with @wheresrhys.

commuterjoy commented:
I talked with @samgiles and @AlbertoElias about a having a shared instrumentation that each origami consumer can use to measure usage/behaviour etc. Eg, o-viewability would be a lot clearer in scope.

AlbertoElias commented:
Ok, let's discuss again at the meeting this week. It was discussed last week and agreed to do it like this, but if @robin-marr-ft is ok with it, let's talk about having a separate module just for this, or for this and maybe other useful features

AlbertoElias commented on main.js:
```diff
@@ -1,129 +1,69 @@
-		broadcast('resize', {
-			viewport: getSize(),
-			originalEvent: ev
+	if (elements.length) {
```
Why is this necessary?

AlbertoElias commented on README.md:
```diff
@@ -1,34 +1,48 @@
+Utility for attaching debounced listeners to resize, scroll, orientation and visibility events on `window`.
 
-*Note: within the module's API and in the documentation below `orientation` is used instead of `orientationchange`, but the actual browser event listened to is `orientationchange`*
+*Note: within the module's API and in the documentation below `orientation` and `visibility` are used instead of `orientationchange` or `visibilitychange`, but the actual browser event listened to is `orientationchange` or `visibilitychange`*
```
should be and instead of or?

AlbertoElias commented on README.md:
```diff
@@ -1,34 +1,48 @@
+Attaches a debounced/throttled (as appropriate) listener to events on window [`resize`, `scroll`, `orientation` or `visibility`] which in turn fires events within the `oViewport` namespace (see **Events** below).
+
+### `o-viewport#trackElements(selector)`
+Tracks the selected elements and report the percentage of the element currently on the screen, changes to the percentage in view fires events within the `oViewport` namespace (see **Events** below).
```
reports

AlbertoElias commented on test/viewport.test.js:
```diff
@@ -1,11 +1,30 @@
+	before(function(){
+		if(isPhantom()) {
+			// hack to make test run in PhantomJS
+			// it doesn't support visibilitychange
```
Maybe add support to it to the polyfill service and document it in the origami.josn

AlbertoElias commented on README.md:
```diff
@@ -1,34 +1,48 @@
+### `o-viewport#trackElements(selector)`
+Tracks the selected elements and report the percentage of the element currently on the screen, changes to the percentage in view fires events within the `oViewport` namespace (see **Events** below).
+
+If the selector argument is omitted `[data-o-viewport-track]` is used.
```
comma after omitted

AlbertoElias commented on README.md:
```diff
@@ -1,34 +1,48 @@
+
+If the selector argument is omitted `[data-o-viewport-track]` is used.
+
+*Note: Enabling element tracking will also attach all other o-viewport events.*
```
Why?

AlbertoElias commented on README.md:
```diff
@@ -1,34 +1,48 @@
-Attaches a debounced/throttled (as appropriate) listener to events on window [`resize`, `scroll` or `orientation`] which in turn fires events within the `oViewport` namespace (see **Events** below).
+Attaches a debounced/throttled (as appropriate) listener to events on window [`resize`, `scroll`, `orientation` or `visibility`] which in turn fires events within the `oViewport` namespace (see **Events** below).
+
+### `o-viewport#trackElements(selector)`
```
selectorS? Is it an array?

AlbertoElias commented on README.md:
```diff
@@ -1,34 +1,48 @@
-Provides a reliable way of obtaining the current dimensions of the viewport. returns an object with the properties `width` and `height`
+Provides a reliable way of obtaining the current dimensions of the browser window. returns an object with the properties `width` and `height`
+
+### `o-viewport#getViewportSize()`
```
Is this a breaking change?

AlbertoElias commented on main.js:
```diff
@@ -1,129 +1,69 @@
+function trackElements(selector){
+	var elements;
+	selector = typeof selector === 'string' ? selector : '[data-o-viewport-track]';
+	try{
```
Is this necessary? If it can't find anything, it will return an empty array

AlbertoElias commented on src/tracking.js:
```diff
@@ -0,0 +1,74 @@
+
+module.exports = {
+	track: Element,
+	tracked: tracked,
```
would call it trackedElements. tracked I thought would be a true-false value

AlbertoElias commented on src/tracking.js:
```diff
@@ -0,0 +1,74 @@
+function update(){
+	tracked.forEach(function (element){
+		if(element.lastResult !== element.percentInViewport()){
+			var inView = element.lastResult = element.percentInViewport();
```
why set a new variable instead of just .lastResult?

AlbertoElias commented:
Should we close this in favour of https://github.com/Financial-Times/o-element-visibility

wheresrhys commented:
:+1:


On 21 May 2015 at 14:18, Alberto Elias <notifications@github.com> wrote:

> Should we close this in favour of
> https://github.com/Financial-Times/o-element-visibility
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Financial-Times/o-viewport/pull/17#issuecomment-104275030>
> .
>

-- 

------------------------------
This email was sent by a company owned by Pearson plc, registered office at 
80 Strand, London WC2R 0RL.  Registered in England and Wales with company 
number 53723.

robin-marr-ft commented:
I've cleaned this up now. I've added a couple of things used by element visibility, namely the visibility event and the getScrollPosition function. The main file was getting a bit long so spilt some methods out into a utils file. Everybody happy if I merge?

wheresrhys commented:
Looks good to me (though it's not my decision)

AlbertoElias commented:
Are things here that it's only use case is in o-element-viewability or that you think should maybe be in there directly because you won't use them if you're not using that module?

robin-marr-ft commented:
Nope, all that stuff has been removed.

AlbertoElias commented:
lgtm then
